File: Solutions/Cpp/move_pieces_to_obtain_a_string.cpp
/*
    [MEDIUM]
    2337. Move Pieces to Obtain a String

    Concepts:
    - string

    Stats:
        Runtime | 39 ms     [Beats 14.37%]
        Memory  | 50.61 MB  [Beats 5.40%]
*/

class Solution {
public:
    bool canChange(string start, string target) {
        vector<pair<int, char>> s_chars = {};
        vector<pair<int, char>> t_chars = {};

        for (int i = 0; i < start.size(); i++) {
            if (start.at(i) == 'L' || start.at(i) == 'R')
                s_chars.push_back(pair(i, start.at(i)));
            if (target.at(i) == 'L' || target.at(i) == 'R')
                t_chars.push_back(pair(i, target.at(i)));
        }

        if (s_chars.size() != t_chars.size()) return false;

        int si, ti;
        char sx, tx;
        for (int i = 0; i < s_chars.size(); i++) {
            auto [si, sx] = s_chars[i];
            auto [ti, tx] = t_chars[i];

            if (sx != tx) return false;

            if (sx == 'L') {
                if (ti > si)
                    return false;
            } else if (sx == 'R') {
                if (ti < si)
                    return false;
            }
        }

        return true;
    }
};

File: Solutions/Python/move_pieces_to_obtain_a_string.py
"""
    [MEDIUM]
    2337. Move Pieces to Obtain a String

    Concepts:
    - string

    Stats:
        Runtime | 195 ms    [Beats 22.58%]
        Memory  | 38.47 MB  [Beats 6.59%]
"""

class Solution:
    def canChange(self, start: str, target: str) -> bool:
        """
            1) store indices of all Ls and Rs
            2) iterate over each char in target, checking
                if next matching char in start is valid
            3) if any discrepancy found, return False
        """
        s_chars = [(i, x) for i, x in enumerate(start) if x in "LR"]
        t_chars = [(i, x) for i, x in enumerate(target) if x in "LR"]

        if len(s_chars) != len(t_chars):
            return False

        for i in range(len(s_chars)):
            si, sx = s_chars[i]
            ti, tx = t_chars[i]

            if sx != tx:
                return False

            if sx == "L":
                if ti > si:
                    return False
            elif sx == "R":
                if ti < si:
                    return False

        return True